src/api/Nhs.Appointments.Core/AvailabilityWriteService.cs

```diff
@@ -0,0 +1,87 @@
+using System.Globalization;
+
+namespace Nhs.Appointments.Core;
+
+public class AvailabilityWriteService(
+    IAvailabilityStore availabilityStore,
+    IAvailabilityCreatedEventStore availabilityCreatedEventStore,
+    IBookingWriteService bookingWriteService) : IAvailabilityWriteService
+{
+    public async Task ApplyAvailabilityTemplateAsync(string site, DateOnly from, DateOnly until, Template template, ApplyAvailabilityMode mode, string user)
+    {
+        if (string.IsNullOrEmpty(site))
+            throw new ArgumentException("site must have a value");
+
+        if (from > until)
+            throw new ArgumentException("until date must be after from date");
+
+        if (template == null)
+            throw new ArgumentException("template must be provided");
+
+        if (template.Sessions is null || template.Sessions.Length == 0)
+            throw new ArgumentException("template must contain one or more sessions");
+
+        if (template.Days is null || template.Days.Length == 0)
+            throw new ArgumentException("template must specify one or more weekdays");
+
+        var dates = GetDatesBetween(from, until, template.Days);
+        foreach (var date in dates)
+        {
+            await availabilityStore.ApplyAvailabilityTemplate(site, date, template.Sessions, mode);
+            await bookingWriteService.RecalculateAppointmentStatuses(site, date);
+        }
+
+        await availabilityCreatedEventStore.LogTemplateCreated(site, from, until, template, user);
+    }
+
+    public async Task ApplySingleDateSessionAsync(DateOnly date, string site, Session[] sessions,
+        ApplyAvailabilityMode mode, string user, Session sessionToEdit = null)
+    {
+        await SetAvailabilityAsync(date, site, sessions, mode, sessionToEdit);
+        await bookingWriteService.RecalculateAppointmentStatuses(site, date);
```

Comment:
I'd be tempted to move the RecalculateAppointmentStatuses call inside SetAvailabilityAsync, because future devs might call SetAvailabilityAsync and forget to do the recalculation.